utilities/basic-utils/src/main/kotlin/NativeMemoryAllocator.kt

```diff
@@ -7,15 +7,30 @@ package org.jetbrains.kotlin.konan.util
 
 import sun.misc.Unsafe
 
-private val allocatorHolder = ThreadLocal<NativeMemoryAllocator>()
-val nativeMemoryAllocator: NativeMemoryAllocator
-    get() = allocatorHolder.get() ?: NativeMemoryAllocator().also { allocatorHolder.set(it) }
+@PublishedApi
+@Volatile
+internal var allocatorHolder: NativeMemoryAllocator? = null
+
+@PublishedApi
+internal var counter = java.util.concurrent.atomic.AtomicInteger(0)
 
-fun disposeNativeMemoryAllocator() {
-    allocatorHolder.get()?.freeAll()
-    allocatorHolder.remove()
+inline fun <R> usingNativeMemoryAllocator(block: () -> R): R {
+    if (counter.getAndIncrement() == 0) {
+        allocatorHolder = NativeMemoryAllocator()
+    }
+    return try {
+        block()
+    } finally {
+        if (counter.decrementAndGet() == 0) {
```

Comment:
What if other thread enters `usingNativeMemoryAllocator` and reaches `block()` while this thread is after the line above (and before the line below)?